### DISCUSSION TITLE
How to pass single topics in a map?

### DISCUSSION THREAD

**Discussion by theletterf** [dita-ot/dita-ot#4485] [Q&A] [answered]
Hi there!

This is a follow-up of https://github.com/orgs/dita-ot/discussions/4484. I feel like I'm so close to the solution, but docs and existing plugins aren't entirely clear on this.

**Context:** 

I want to pass each DITA topic in a map to the Vale linter using an <exec> statement in the Ant script. For that, I need a variable that allows me to iterate through each file in the ditamap I'm providing.

**Command:**

```sh
dita --input=sample.ditamap --format=html5
```

**Plugin.xml:**

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="dita-ot/plugin.rnc" type="application/relax-ng-compact-syntax"?>
<plugin id="org.splunk.vale.check" version="1.0">
   <feature extension="package.support.name" value="Fabrizio Ferri Benedetti"/>
   <feature extension="package.support.email" value="fferribenedetti@splunk.com"/>
   <feature extension="package.version" value="1.0"/>
   <feature extension="ant.import" file="vale.xml"/>
   <feature extension="depend.preprocess.post" value="vale-check"/>
</plugin>
```

**vale.xml:**

```xml
<?xml version='1.0' encoding='UTF-8'?>
<project name="vale-check" default="vale-check" basedir="." xmlns:artifact="antlib:org.apache.maven.artifact.ant">
  <target name="vale-check">

      <ditafileset format="dita"/>
      <echo>Checking: </echo>
      <echo>${dita.input.filename}</echo>
      <exec executable="vale" failonerror="true" spawn="false" logError="true" outputproperty="cmdOutput" errorproperty="cmdError" dir="${dita.temp.dir}">
      <arg value="${dita.input.filename}" />
      </exec>
      <!-- Print the standard output to the console -->
      <echo>Standard Output:</echo>
      <echo>${cmdOutput}</echo>
      <!-- Print the error output to the console, if any -->
      <echo>Standard Error:</echo>
      <echo>${cmdError}</echo>
</target>
</project>
```

The current out I'm getting shows that the script is processing the ditamap itself, not each topic:

```sh
Checking:
sample.ditamap
Standard Output:

 sample.ditamap
 1:109  warning  Use American-style formatting   Splunk.Decimals
                 for commas and decimal points
                 in numbers.
 1:200  warning  Use American-style formatting   Splunk.Decimals
                 for commas and decimal points
                 in numbers.

✖ 0 errors, 2 warnings and 0 suggestions in 1 file.
Standard Error:
```

**Comment by raducoravu** [comment]
On what files do you want to apply the vale executable:

- On the original DITA XML files from the project input folder
- On the pre-processed DITA XML files from the temporary files folder which have all the content references and key references resolved
- On the output HTML files 

?
I see vale has some support for directly processing DITA XML files using the DITA Open Toolkit:
https://vale.sh/docs/topics/scoping#dita
Have you tried this support? Does it properly report errors with a degree of accuracy that it is easy to locate them back in the original DITA XML content?
Because if for example you apply vale on the output HTML content and obtain some errors with line and column number, they might not be that easy to map in the original DITA XML content which has DOCTYPE declaration and the conrefs and keyrefs are not expanded.

**Reply by theletterf** [reply]
@raducoravu Thanks so much for reaching out!

DITA processing in Vale uses DITA OT and is quite accurate. When I run it on a single file the position of the offending strings is accurate.

In this sense, I'd need to preserve the DITA source as intact as possible when applying Vale. When I run the script, Vale is supposedly processing the ditamap, but the outcome doesn't make sense, which leads me to believe that it's checking something else entirely.

Is DITA OT capable of providing the path to the original DITA XML files, or at least the pre-processed DITA XML files so that the output can still reference what files contain which errors?

**Reply by theletterf** [reply]
Reading about https://www.dita-ot.org/dev/reference/preprocess-genlist I wonder if those are resources I could access and leverage at preprocess phase. I don't know how, though. @infotexture Any hint as to what approach would be better here?

**Reply by raducoravu** [reply]
In my opinion you do not need a DITA OT plugin, you just need to create a script which takes as parameter the folder containing all the DITA topics and maps, iterates them and runs "vale" over each of them. The DITA OT is mostly a publishing engine which transforms content (DITA to HTML, DITA to X, DITA to Y). But in your case you do not seem to want to process the DITA XML content in any way.

**Reply by theletterf** [reply]
I agree, but let me provide a bit more context: I aim to use this plugin in the context of a DITA-OT capable CCMS (Heretto). Writers would be able to export a map to the plugin to get a QA check.

**Reply by theletterf** [reply]
@infotexture @raducoravu I'd still like to try pass files to Vale through the plugin. I can't figure out what the proper way is, though, because besides using `<ditafileset format="dita"/>` I don't know what predefined variables are at my disposal for passing the dita topics as files.

**Reply by raducoravu** [reply, marked as answer]
The DITA OT is a publishing engine which processes DITA XML content, it transforms the DITA XML content to various outputs, this is its main purpose, so its main purpose is not to apply an external process on each of the original DITA files. This processing is usually done with Java pre-processing steps and XSLT steps which can use the "ditafileset format="dita"". Again, you do not need that, the ditafileset format="dita" returns the pre-processed DITA XML files from the temporary files folder with all content references expanded, keys resolved, which again is not something you want.
Maybe the closest plugin I created similar to what you want is this one:
https://github.com/oxygenxml/dita-export-map/tree/master/com.oxygenxml.export.map
It copies the original DITA Map and topics to a separate folder and then zips them. You could copy them to a folder and then run the vale executable over each file in that folder.

**Reply by theletterf** [reply]
Thanks, I'll try that!
